tangelo/linq/simulator.py

```diff
@@ -26,515 +26,64 @@
-            else:
-                job_sim = cirq_simulator.simulate(translated_circuit, initial_state=cirq_initial_statevector)
-                self._current_state = job_sim.final_state_vector
-                frequencies = self._statevector_to_frequencies(self._current_state)
-
-            return (frequencies, np.array(self._current_state)) if return_statevector else (frequencies, None)
-
-    def get_expectation_value(self, qubit_operator, state_prep_circuit, initial_statevector=None):
-        r"""Take as input a qubit operator H and a quantum circuit preparing a
-        state |\psi>. Return the expectation value <\psi | H | \psi>.
-
-        In the case of a noiseless simulation, if the target backend exposes the
-        statevector then it is used directly to compute expectation values, or
-        draw samples if required. In the case of a noisy simulator, or if the
-        statevector is not available on the target backend, individual shots
-        must be run and the workflow is akin to what we would expect from an
-        actual QPU.
-
-        Args:
-            qubit_operator(openfermion-style QubitOperator class): a qubit
-                operator.
-            state_prep_circuit: an abstract circuit used for state preparation.
-
-        Returns:
-            complex: The expectation value of this operator with regards to the
-                state preparation.
-        """
-        # Check if simulator supports statevector
-        if initial_statevector is not None and not self.statevector_available:
-            raise ValueError(f'Statevector not supported in {self._target}')
-
-        # Check that qubit operator does not operate on qubits beyond circuit size.
-        # Keep track if coefficients are real or not
-        are_coefficients_real = True
-        for term, coef in qubit_operator.terms.items():
-            if state_prep_circuit.width < len(term):
-                raise ValueError(f'Term {term} requires more qubits than the circuit contains ({state_prep_circuit.width})')
-            if type(coef) in {complex, np.complex64, np.complex128}:
-                are_coefficients_real = False
 
-        # If the underlying operator is hermitian, expectation value is real and can be computed right away
-        if are_coefficients_real:
-            if self._noise_model or not self.statevector_available \
-                    or state_prep_circuit.is_mixed_state or state_prep_circuit.size == 0:
-                return self._get_expectation_value_from_frequencies(qubit_operator, state_prep_circuit, initial_statevector=initial_statevector)
-            elif self.statevector_available:
-                return self._get_expectation_value_from_statevector(qubit_operator, state_prep_circuit, initial_statevector=initial_statevector)
-
-        # Else, separate the operator into 2 hermitian operators, use linearity and call this function twice
+        # Return new instance of target class
+        if target is None:
```

Comment:
I don't think L58-59 should exist anymore.

Your method signature handles a default value, neither your documentation nor your error message below mentions `None` as a supported value.

With your default value, users would have to go out of their way to stipulate `target=None` to be in this situation, which seems non-sensical to me.

Reply:
Unfortunately, all our algorithms `VQESolver, ADAPTSolver,...` that use `Simulator` send `None` as the default `target`. I could change that behaviour in this PR but that adds many changed files to this PR.

Reply:
Thank you ! I think this could be a separate PR then.
For now, if you are supporting `None` as a value for target, your current documentation and error message should however be consistent.